### ISSUE TITLE
[HOLD for payment][Bug] After typing even a mildly long comment in the Android compose box, the whole app starts to hang

### ISSUE THREAD

**Issue by mvtglobally** [Expensify/App#11321]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open a chat -- even an empty one
2. Just start typing a ton on the keyboard.  Not copy/pasting, just finger-smash the keyboard for about 30s

## Expected Result:
App should respond quickly 

## Actual Result:
After typing even a mildly long comment in the Android compose box, the whole app starts to hang.

## Workaround:
unknown 

## Platform:

Where is this issue occurring?

- Android

**Version Number:** 1.2.3-0
**Reproducible in staging?:** 
**Reproducible in production?:** 
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation 
![image - 2022-09-27T002415 487](https://user-images.githubusercontent.com/43995119/192431597-1ae78277-cb91-4131-891c-b40aa7fe0143.png)

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @JmillsExpensify @quinthar 
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1663172358453739

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by Beamanator** [comment]
Hmm I believe this should be able to be investigated by contributors so I'll make it `External`, but at the moment I'm having trouble reproducing - I logged in on my Android device, mashed the keyboard for a long time, didn't see any slowdown 🤔  It seems this might have to happen in a HT account, even thought it seems like @mvtglobally tried testing on a HT account already?

**Comment by kadiealexander** [comment]
@Beamanator I was able to repro this on an asus phone with android 12, but it took quite a long message (40k characters).

**Comment by kadiealexander** [comment]
https://www.upwork.com/jobs/~01ada9667fd1d8d5a9
https://www.upwork.com/ab/applicants/1576766472536219648/job-details

**Comment by quinthar** [comment]
I am so confused, this is incredibly easy to reproduce with 100% consistency on my Pixel 5 within about 10 seconds.  Just:
1. Open the app and sign into an account (or create a new one)
2. Open any chat, even a new one
3. Just drum your fingers over the keyboard repeatedly
4. After about 10s it should hang
5. If that doesn't work, just find a long message, and hold down backspace -- then it'll quickly hang

Are you saying that those steps do not reproduce it for you?

**Comment by Beamanator** [comment]
Hmm I'm on a Samsung Galaxy A72 and I followed your exact steps, didn't see anything hang or really slow down at all 🤷

**Comment by quinthar** [comment]
Just did a fresh uninstall/reinstall, created a new account, and was able to reproduce in about 15 seconds:

![image](https://user-images.githubusercontent.com/575186/193653976-f10d663f-f463-4c6d-ade5-0af26aeb6aa6.png)

The keyboard is still working in that I am still able to press keys, but the input box is hung and no new input actually goes into it.  I suspect it has something to do with saving the draft too quickly?  Do we batch changes to the draft, or save fresh with every keystroke?

**Comment by Beamanator** [comment]
Interesting, I was able to reproduce on a Pixel 4 simulator! Though after mashing the keyboard of my Samsung Galaxy A72 for literally a minute, I still didn't see any slowdown 🤷 

As for saving the draft to Onyx, it looks like we do debounce here: https://github.com/Expensify/App/blob/1a42cf8a48bcb51b7b469f342a3f122de7aaef34/src/pages/home/report/ReportActionCompose.js#L117

That ^ debounce looks like it's working properly, so I'm not sure what else would be causing this, I think this should be part of the WhatsApp Quality project so I'll add it there

**Comment by Beamanator** [comment]
Waiting on proposals, I think we can double this to make it more enticing @kadiealexander

**Comment by kadiealexander** [comment]
Price doubled!!

**Comment by JmillsExpensify** [comment]
Aren't we already working on resolving this issue with [this PR](https://github.com/Expensify/App/pull/11684)?

**Comment by mallenexpensify** [comment]
@sobitneupane , do you think the fix for this will also fix https://github.com/Expensify/App/issues/11954?

**Comment by sobitneupane** [comment]
Both the issues are related to long text and looks like we already have a [ PR](https://github.com/Expensify/App/pull/11684) up for improving input performance. So, I suggest holding both the issues.

cc: @Beamanator @mallenexpensify

**Comment by hannojg** [comment]
Hey guys, I have already tested with @quinthar manually if https://github.com/Expensify/App/pull/11684 fixes this issue. Unfortunately, it doesn't. 

Up until today i wasn't able to reproduce that issue myself. @Beamanator you said you were able to reproduce the issue with a Pixel 4 simulator, that sounds great. Can you share the specs?

I really would like to fix this bug, but def need a reproduction 👍

**Comment by Beamanator** [comment]
Hey @hannojg ! Yes I'll re-test on the Pixel 4 simulator and keep you posted :D

> Can you share the specs/ .ini file of your emulator?

Hmm how can I find this file? 😅

**Comment by Beamanator** [comment]
Looks like Github doesn't support those `.ini` file format, I can upload to Google Drive or Slack

**Comment by hannojg** [comment]
@Beamanator Usually your emulators are stored in your home directory under the following path:

`.android/avd/YOUR_EMULATOR_NAME.avd/config.ini`

in addition to that your `.android/avd/YOUR_EMULATOR_NAME.ini` file would be interesting!

> I can upload to Google Drive or Slack

Oh yeah sure, or you just create a code comment with \`\`\` and post it here as text 😊

**Comment by Beamanator** [comment]
Note: On the simulator, after smashing the keyboard for a while and experiencing the app pretty much hanging, if I close the report, navigate to another, then navigate back, it didn't seem slow anymore. The video from my test didn't save 🙃 so I'll try again in a bit

Here's the `config.ini` file:

```
AvdId = Pixel4
PlayStore.enabled = true
abi.type = arm64-v8a
avd.ini.displayname = Pixel4
avd.ini.encoding = UTF-8
disk.dataPartition.size = 6442450944
fastboot.chosenSnapshotFile = 
fastboot.forceChosenSnapshotBoot = no
fastboot.forceColdBoot = no
fastboot.forceFastBoot = yes
hw.accelerometer = yes
hw.arc = false
hw.audioInput = yes
hw.battery = yes
hw.camera.back = virtualscene
hw.camera.front = emulated
hw.cpu.arch = arm64
hw.cpu.ncore = 4
hw.dPad = no
hw.device.hash2 = MD5:6b5943207fe196d842659d2e43022e20
hw.device.manufacturer = Google
hw.device.name = pixel_4
hw.gps = yes
hw.gpu.enabled = yes
hw.gpu.mode = auto
hw.initialOrientation = Portrait
hw.keyboard = yes
hw.lcd.density = 440
hw.lcd.height = 2280
hw.lcd.width = 1080
hw.mainKeys = no
hw.ramSize = 1536
hw.sdCard = yes
hw.sensors.orientation = yes
hw.sensors.proximity = yes
hw.trackBall = no
image.sysdir.1 = system-images/android-33/google_apis_playstore/arm64-v8a/
runtime.network.latency = none
runtime.network.speed = full
sdcard.size = 512M
showDeviceFrame = yes
skin.dynamic = yes
skin.name = pixel_4
skin.path = /Users/alexbeaman/Library/Android/sdk/skins/pixel_4
tag.display = Google Play
tag.id = google_apis_playstore
vm.heapSize = 256
```

**Comment by Beamanator** [comment]
And here's my `Pixel4.ini` file:

```
avd.ini.encoding=UTF-8
path=/Users/alexbeaman/.android/avd/Pixel4.avd
path.rel=avd/Pixel4.avd
target=android-33

```

**Comment by hannojg** [comment]
Thx! may i ask on which OS your computer is on? (if its macOS, are you running on an M1?)

**Comment by Beamanator** [comment]
https://user-images.githubusercontent.com/3885503/196717445-590c1d88-8c87-4076-9b2a-38fcd9df1d4c.mov

> may i ask on which OS your computer is on? (if its macOS, are you running on an M1?)

Sure thing, yes I'm on a macOS (12.6), Chip: M1 Max

**Comment by quinthar** [comment]
Oh great yes, that's exactly what I'm experiencing!

**Comment by quinthar** [comment]
@hannojg  any luck?  Excited for this one!

**Comment by hannojg** [comment]
Not yet, will work on it this week, will keep y'all updated 😊

**Comment by techievivek** [comment]
Note: Something similar happened [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1666892266003079)(but the chat count is extremely high 51k so I suppose this is expected) but it will be good to counter all the platforms and not just Android.

**Comment by Beamanator** [comment]
Assigning @hannojg so your comments also clear the `Overdue` status :D

**Comment by kadiealexander** [comment]
@hannojg don't forget to apply on Upwork so we can issue payment when it's due! :)

**Comment by quinthar** [comment]
Any luck with this?  This issue makes the app entirely unusable for me.

**Comment by hannojg** [comment]
Hey, I spend some more hours on this. I can exclude now more parts that aren't causing the issue, but still can't tell which one it is exactly. 

It's a bit hard to estimate, but I think after ~5 more hours I should find the bug.

Do you want me/us to continue investigating in that issue (asking because it becomes a bit of a time sink) @quinthar ?

**Comment by quinthar** [comment]
Yes please do. Thanks!

**Comment by mallenexpensify** [comment]
@hannojg any update? Would love to get this fixed for ~DB~ Android users, there are also other issues held on this fix.  Bumping to `Daily`

**Comment by hannojg** [comment]
Haven't continued yet, you can expect results latest by the end of this week!

**Comment by quinthar** [comment]
Haha, thank you.  I'd love to be able to, you know, use our own app.

**Comment by quinthar** [comment]
Awesome!

**Comment by JmillsExpensify** [comment]
Nice one! @quinthar have you had a chance to take the APK for a spin yet? Does that resolve your issue?

**Comment by quinthar** [comment]
I have not, can you send me a link?

**Comment by JmillsExpensify** [comment]
Ah ya, it's here: https://drive.google.com/file/d/1HvRg_d97D-R0YlGD49MEqr-ByC9VYBXU/view?usp=sharing

**Comment by quinthar** [comment]
Wooohooo!!!!  Seems to work!!!!!!! (but shows messages in the wrong order)

**Comment by Beamanator** [comment]
@hannojg is making great progress! 👍 (just updating Melvin)

**Comment by Beamanator** [comment]
Just updated this issue, so not overdue Melvin 🙃

**Comment by strepanier03** [comment]
Latest on this in Slack [HERE](https://expensify.slack.com/archives/C01GTK53T8Q/p1667868583336109)

<img width="410" alt="2022-11-09_12-15-35" src="https://user-images.githubusercontent.com/10925636/200932797-b2430c8e-e0fa-4603-a635-92693c676b2a.png">

**Comment by kadiealexander** [comment]
See above, not overdue.

**Comment by hannojg** [comment]
Hey guys, going to continue looking for a fix today. So we just know the cause of the issue, but don't have a fix yet.
Working on it 😼 😊 !

**Comment by Beamanator** [comment]
@hannojg still working on a fix! 👍

**Comment by JmillsExpensify** [comment]
On that note, @hannojg it would be good to circle back on the plan from here. I don't see a PR linked to this issue, are you working on that somewhere else?

**Comment by hannojg** [comment]
Hey, I am still nailing down a fix. The issue is present in react-native (so not specific to expensify code usage), and it might be an android bug, so I will have to create an issue ticket. Will link PRs and external issues as soon as I am there, but so far there are no code changes to expensify that I could share. Still working on it and will provide results today! 😊

**Comment by hannojg** [comment]
Here is the react native issue: https://github.com/facebook/react-native/issues/35350

Will now check whether this is to fix in react native's code or whether its an android bug.

**Comment by hannojg** [comment]
Update: I am still looking for a fix within the react-native source code. I don't think it's an android native bug, as I am unable to reproduce the issue just using android native.

**Comment by hannojg** [comment]
Okay, I have a proposal available for a _workaround_ until I get to fix the issue upstream at react native.
The workaround is a bit ugly, involves changing the react-native source code … BUT I am a bit uncertain how long it will take to get the issue fixed upstream. 
Until then I think the workaround will be valuable enough (android 13 users can't really use the app rn).

Will put up a PR in a moment …

**Comment by quinthar** [comment]
It's 100% fine to fix our fork of RN; we don't really care if it gets
merged upstream.  We would always prefer to fix the bug where it actually
happens, even if it's never merged into mainline.

**Comment by hannojg** [comment]
Yes, a 100%, it's just that the workaround isn't a fix. For the fix itself, I have to dive deep into react native's code. However, for the workaround, I still need to modify some react-native code - so that's only temporarily until we have a proper fix.

**Comment by quinthar** [comment]
to clarify, do you have an actual fix?  Why do the workaround?

**Comment by hannojg** [comment]
No, i don't have a proper fix yet.
We can also wait until I have a proper fix.

**Comment by quinthar** [comment]
Oh if we don't have a fix, then yes, let's do the workaround.

**Comment by hannojg** [comment]
A draft is up for circumventing the issue until a proper fix is implemented:
https://github.com/Expensify/App/pull/12820

**Comment by quinthar** [comment]
Can you describe the underlying problem at a high level, as well as how the
workaround works?

**Comment by hannojg** [comment]
## The underlying problem

To achieve the inverted list effect react native simply flips the `FlatList` vertically with `scaleY: -1`. It then also flips all items with the same transformation.
On the native side this translates to the following transformation commands, which get applied to the view:

- transformX: 0
- transformY: 0
- rotate: 180
- rotateY: 0
- rotateX: 180 (now the view looks as usual again as we reverted the previous rotation)
- scaleX: 0
- scaleY: -1

The problem now is, that on android 13 a bug was introduced that causes ANRs when `rotateX` and `scaleY` are set and we simultaneously render stuff (animations, entering text). It's a bit hard to tell exactly what's going wrong on the android side, but that's also nothing we can really help with. 

**I am currently in the process of filing a bug report to the android OSS project. Will link soon.**

## How the workaround works

We manually set the styles for the FlatList and its items to achieve the inverted effect. However, we add a `scaleX: -1`, which changes the transform operations on the native side to skip `rotateX: 180` (it's then only `rotateY: -180`). This way we circumvent the bug.

When doing this the only issue is, that as we have mirrored the FlatList on the X-Axis the ScrollBar will be on the wrong side. That's why I added a prop called `verticalScrollbarPosition` to the ScrollView, so we can change its position.

## Long term

I think it would be best to get the issue fixed on android and then revert the workaround. Another solution could be to modify how react native applies transformations, something I want to look into as well (its a bit redundant to apply transformations that are not really needed).

**Comment by quinthar** [comment]
Wow, very interesting!

**Comment by quinthar** [comment]
I'm curious if you have an ETA for when this workaround can be merged?

**Comment by JmillsExpensify** [comment]
Closing the loop. [PR is here](https://github.com/Expensify/App/pull/12820). Discussion also happening [in slack](https://expensify.slack.com/archives/C035J5C9FAP/p1669022791923689).

**Comment by Beamanator** [comment]
Yep 👍  PR is approved, just need to get a quick change to our `react-native` fork published, then we can incorporate that change in `App` 👍

**Comment by mountiny** [comment]
The PR has been merged 🎉  lets wait to see how this will perform in staging

**Comment by JmillsExpensify** [comment]
Looks like this is on production at this point. @quinthar Has mobile performance improved for this case?

**Comment by quinthar** [comment]
yes, this seems to be fixed, it's miraculous, thank you!

**Comment by JmillsExpensify** [comment]
Woo! Go team. @kadiealexander We should make sure that contributors and related are paid out before closing the issue. Thanks!

**Comment by kadiealexander** [comment]
https://expensify.slack.com/archives/C01SKUP7QR0/p1669595062240009

**Comment by kadiealexander** [comment]
The old Upwork job closed, so please apply here @hannojg and @sobitneupane and I'll issue payment for you both. :)

https://www.upwork.com/jobs/~015cde3e5ee4ee738f
https://www.upwork.com/ab/applicants/1597366602525372416/job-details

**Comment by hannojg** [comment]
Sorry @kadiealexander, but I can't open any of those links 😅 can you help me get access to it? (I am part of #expensify-margelo and #expensify-open-source on slack)

**Comment by kadiealexander** [comment]
Ah! To receive this bonus you may need to set up a profile in Upwork, which is how we typically pay our freelance contributors.

**Comment by sobitneupane** [comment]
@kadiealexander I didn't review the proposal and PR. So, I am not eligible for payment.

**Comment by JmillsExpensify** [comment]
@adelekennedy I think we have some confusion here, sorry if I caused it. @hannojg is working with on a special engagement. And then it sounds like @sobitneupane did not review the linked PR, so when I say "make sure" I mainly meant let's double check we're all good here before closing.

**Comment by kadiealexander** [comment]
@hannojg after internal discussion, you're eligible for the $500 bonus for this issue if you wish. You'll just need to create an Upwork profile to process the payment. Please let me know how you wish to proceed!

**Comment by kadiealexander** [comment]
Discussing payment delivery [here](https://expensify.slack.com/archives/C035J5C9FAP/p1669802801418749).

**Comment by hannojg** [comment]
Payment is handled separately through a invoice by Margelo!

**Comment by mountiny** [comment]
In that case we can close this one, nice job @hannojg!

**Comment by justinfauci** [comment]
@hannojg Thank you so much for your work on this, it solved a major problem we were having with our messenger for Android 13 devices.
